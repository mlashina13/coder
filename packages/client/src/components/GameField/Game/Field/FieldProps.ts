export type FieldProps = {
  /** Объект canvas */
  canvas: HTMLCanvasElement;
  /** Контекст canvas */
  ctx: CanvasRenderingContext2D;
  /** Размер игровой фишки */
  chipSize: number;
  /** Количество цветов в последовательности */
  colorsInRowCount: number;
  /** Максимальное количество ходов */
  maxStepsCount: number;
  /** Координата источника освещения по оси X */
  lightX: number;
  /** Координата источника освещения по оси Y */
  lightY: number;
};
