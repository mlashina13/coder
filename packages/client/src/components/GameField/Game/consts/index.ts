export const enum chipColors {
  red = '#7b001c',
  yellow = '#fedd00',
  green = '#0f3325',
  blue = '#191970',
  purple = '#4b0082',
  orange = '#ff8c00',
  pink = '#c71585',
  brown = '#800000',
  gray = '#696969',
  cyan = '#008080',
  olive = '#556b2F',
}

export const lighterChipColors = {
  [chipColors.red]: '#ff6666',
  [chipColors.yellow]: '#ffff66',
  [chipColors.green]: '#66cc66',
  [chipColors.blue]: '#6666ff',
  [chipColors.purple]: '#cc66cc',
  [chipColors.orange]: '#ffcc66',
  [chipColors.pink]: '#ffe6f2',
  [chipColors.brown]: '#e69999',
  [chipColors.gray]: '#cccccc',
  [chipColors.cyan]: '#99ffff',
  [chipColors.olive]: '#b3b366',
};

export const gradientOffsetX = -3;
export const gradientOffsetY = -2;
export const gradientRadius = 2;

export const enum fieldColors {
  white = 'white',
  black = 'clack',
  lightSlateGray = 'lightSlateGray',
}

export const availableColors = [
  chipColors.red,
  chipColors.yellow,
  chipColors.green,
  chipColors.blue,
  chipColors.purple,
  chipColors.orange,
  chipColors.pink,
  chipColors.brown,
  chipColors.gray,
  chipColors.cyan,
  chipColors.olive,
];

export const backgroundColor = 'lightSlateGray' as chipColors;
export const chipSize = 25;
export const checkChipRadius = 5;

export const distanceBetweenChips = 25;

export const leftMouseButton = 0;

export const startX = 50;
export const startY = 50;

const millisecond = 1;
export const second = 1000 * millisecond;
export const minute = 60 * second;
