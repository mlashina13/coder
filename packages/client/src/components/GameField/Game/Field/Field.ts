import { backgroundColor, chipSize } from '../consts';

export default class Field {
  private static _instance: Field | void;

  /** Элемент canvas */
  private readonly _canvas!: HTMLCanvasElement;

  /** Контекст canvas */
  private readonly _ctx!: CanvasRenderingContext2D;

  /** Ширина поля с игровыми фишками */
  private readonly _gameChipsFieldWidth!: number;

  /** Высота поля с игровыми фишками */
  private readonly _gameChipsFieldHeight!: number;

  /** Ширина поля с ячейками */
  private readonly _chipSlotsFieldWidth!: number;

  /** Высота поля с ячейками */
  private readonly _chipSlotsFieldHeight!: number;

  /** Ширина поля с фишками проверки */
  private readonly _checkChipsFieldWidth!: number;

  /** Стартовая координата по оси Y поля с ячейками */
  private readonly _chipSlotsFieldStartY!: number;

  /** Стартовая координата по оси X поля с фишками проверки */
  private readonly _checkChipsFieldStartX!: number;

  constructor(
    canvas: HTMLCanvasElement,
    ctx: CanvasRenderingContext2D,
    colorsCount: number,
    stepsCount: number,
    allAvailableColorsCount: number,
    colorsInRowCount: number,
    maxStepsCount: number
  ) {
    if (Field._instance) {
      return Field._instance;
    }

    this._canvas = canvas;

    this._ctx = ctx;
    this._gameChipsFieldWidth =
      chipSize + 2 * colorsCount * chipSize + chipSize * Math.ceil(colorsCount / 2);
    this._gameChipsFieldHeight = chipSize * 3;
    this._chipSlotsFieldWidth = chipSize * (2 * colorsCount + 1);
    this._chipSlotsFieldHeight = chipSize * (2 * stepsCount + 1);
    this._checkChipsFieldWidth = chipSize * Math.ceil(colorsCount / 2);
    this._chipSlotsFieldStartY = chipSize * 5;
    this._checkChipsFieldStartX = chipSize + this._chipSlotsFieldWidth;

    Field._instance = this;

    this.setFieldSize(allAvailableColorsCount, colorsInRowCount, maxStepsCount);
  }

  public get canvas() {
    return this._canvas;
  }

  public get ctx() {
    return this._ctx;
  }

  public get gameChipsFieldWidth() {
    return this._gameChipsFieldWidth;
  }

  /** Установка размера игрового поля */
  private setFieldSize = (
    allAvailableColorsCount: number,
    colorsInRowCount: number,
    maxStepsCount: number
  ) => {
    this._canvas.width = chipSize * (allAvailableColorsCount * 2 + colorsInRowCount / 2 + 1);
    this._canvas.height = chipSize * (7 + 2 * maxStepsCount);
  };

  /** Отрисовка игрового поля */
  public draw = () => {
    this.fill();
    this.drawChipsField();
    this.drawSlotsField();
    this.drawChecksField();
  };

  /** Заполнение фона */
  private fill = () => {
    this._ctx.fillStyle = backgroundColor;

    this._ctx.fillRect(0, 0, this._canvas.width, this._canvas.height);
    this._ctx.strokeRect(0, 0, this._canvas.width, this._canvas.height);
  };

  /** Отрисовка поля для игровых фишек */
  private drawChipsField = () => {
    this._ctx.strokeRect(chipSize, chipSize, this._gameChipsFieldWidth, this._gameChipsFieldHeight);
  };

  /** Отрисовка поля для ячеек */
  private drawSlotsField = () => {
    this._ctx.strokeRect(
      chipSize,
      this._chipSlotsFieldStartY,
      this._chipSlotsFieldWidth,
      this._chipSlotsFieldHeight
    );
  };

  /** Отрисовка поля для фишек проверки */
  private drawChecksField = () => {
    this._ctx.strokeRect(
      this._checkChipsFieldStartX,
      this._chipSlotsFieldStartY,
      this._checkChipsFieldWidth,
      this._chipSlotsFieldHeight
    );
  };

  /** Очищение игрового поля */
  public clearGameField = () => {
    this._ctx.clearRect(0, 0, this._canvas.width, this._canvas.height);
  };

  /** Деинициализация игрового поля */
  public static destructor = () => {
    Field._instance = undefined;
  };
}
