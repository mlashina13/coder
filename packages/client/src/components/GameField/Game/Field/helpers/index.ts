import { gradientOffsetX, gradientOffsetY } from '../../consts';

export const getFieldShadow = (
  x: number,
  y: number,
  lightX: number,
  lightY: number,
  gameChipsFieldWidth: number,
  gameChipsFieldHeight: number
) => {
  const angleRadians = Math.atan2(y - lightY, x - lightX);
  const distanceToLight = Math.sqrt((lightX - x) ** 2 + (lightY - y) ** 2);
  const gradientRadius = distanceToLight / 10;
  const longestSide = Math.max(gameChipsFieldWidth, gameChipsFieldHeight) / gradientRadius;
  const offsetX = Math.cos(angleRadians) * gradientOffsetX;
  const offsetY = Math.sin(angleRadians) * gradientOffsetY;

  return {
    shadowBlur: Math.min(gradientRadius, distanceToLight / (5 * gradientRadius)) * longestSide,
    shadowOffsetX: (offsetX * gameChipsFieldWidth) / (2 * gradientRadius),
    shadowOffsetY: (offsetY * gameChipsFieldHeight) / (2 * gradientRadius),
  };
};
