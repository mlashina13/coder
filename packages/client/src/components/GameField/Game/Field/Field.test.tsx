/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import { setupJestCanvasMock } from 'jest-canvas-mock';
import Field from './Field';
import { FieldProps } from './FieldProps';
import { lightX, lightY } from '../consts';

describe('Field', () => {
  beforeAll(() => {
    setupJestCanvasMock();
  });

  const getCanvas = () => document.createElement('canvas');

  const getCtx = (canvas: HTMLCanvasElement) => {
    const context = canvas.getContext('2d');
    return context as CanvasRenderingContext2D;
  };

  describe('Создание класса', () => {
    const getDefaultProps = () => {
      const canvas = getCanvas();
      const ctx = getCtx(canvas);

      return {
        canvas,
        ctx,
        chipSize: 21.5,
        colorsInRowCount: 4,
        defaultLightX: lightX,
        defaultLightY: lightY,
        maxStepsCount: 1,
      };
    };

    const getDefaultfieldClass = () => {
      const {
        canvas,
        chipSize,
        colorsInRowCount,
        ctx,
        defaultLightX,
        defaultLightY,
        maxStepsCount,
      } = getDefaultProps();
      return new Field(
        canvas,
        ctx,
        chipSize,
        colorsInRowCount,
        maxStepsCount,
        defaultLightX,
        defaultLightY
      );
    };
    test('Должен создавать идентичные классы при одинаковом наборе параметров', () => {
      const firstField = getDefaultfieldClass();
      const secondField = getDefaultfieldClass();

      expect(firstField).toEqual(secondField);
    });

    describe('Геттеры', () => {
      test('canvas не должен быть null', () => {
        const fieldClass = getDefaultfieldClass();
        expect(fieldClass.canvas).not.toBeNull();
      });
      test('ctx не должен быть null', () => {
        const fieldClass = getDefaultfieldClass();
        expect(fieldClass.ctx).not.toBeNull();
      });
      test('gameChipsFieldWidth не должен быть null', () => {
        const fieldClass = getDefaultfieldClass();
        expect(fieldClass.gameChipsFieldWidth).not.toBeNull();
      });
      test('gameChipsFieldWidth должен высчитыватся по формуле', () => {
        const fieldClass = getDefaultfieldClass();
        const { chipSize, colorsInRowCount } = getDefaultProps();
        const checkChipsFieldWidthips = chipSize * (Math.ceil(colorsInRowCount / 2) + 1);
        const chipSlotsFieldWidth = 1.5 * chipSize * (colorsInRowCount + 1);
        const defaultGameChipsFieldWidth = checkChipsFieldWidthips + chipSlotsFieldWidth;
        expect(fieldClass.gameChipsFieldWidth).toEqual(defaultGameChipsFieldWidth);
      });
    });

    test('Метод draw должен отрисовывать поле', () => {
      const fieldClass = getDefaultfieldClass();

      expect(fieldClass.ctx.__getPath()).toMatchSnapshot();
    });
  });
});
