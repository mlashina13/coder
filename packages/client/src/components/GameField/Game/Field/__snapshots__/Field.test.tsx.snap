// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование ф-сти игры: класс Field Создание класса Метод draw должен отрисовывать поле 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
]
`;
